xcdat/temporal_avg.py

```diff
@@ -0,0 +1,988 @@
+        grouping on multiple datetime parameters such as year and month,
+        which xarray does not support. After grouping, the parameters passed to
+        the averaging operation are stored in the data variable's dictionary of
+        attributes.
+
+        Parameters
+        ----------
+        data_var : xr.DataArray
+            The data variable.
+
+        Returns
+        -------
+        xr.DataArray
+            The averaged data variable.
+        """
+        dv = data_var.copy()
+
+        if (
+            self._freq == "season"
+            and self._season_config.get("dec_mode") == "DJF"
+            and self._season_config.get("drop_incomplete_djf") is True
+        ):
+            dv = self._drop_incomplete_djf(dv)
+
+        self._multiindex, self._multiindex_name = self._get_multiindex(dv)
+
+        if self._weighted:
+            self._weights = self._get_weights(dv)
+            dv *= self._weights
+            dv = self._groupby_multiindex(dv).sum()  # type: ignore
+        else:
+            dv = self._groupby_multiindex(dv).mean()  # type: ignore
+
+        dv = self._add_operation_attrs(dv)
+        return dv
+
+    def _group_observations(self, data_var: xr.DataArray) -> DataArrayGroupBy:
+        """Groups the data variable observations to calculate departures.
+
+        Parameters
+        ----------
+        data_var : xr.DataArray
+            The data variable.
+
+        Returns
+        -------
+        DataArrayGroupBy
+            The grouped data variable.
+        """
+        dv = data_var.copy()
+
+        if (
+            self._freq == "season"
+            and self._season_config.get("dec_mode") == "DJF"
+            and self._season_config.get("drop_incomplete_djf") is True
+        ):
+            dv = self._drop_incomplete_djf(dv)
+
+        self._multiindex, self._multiindex_name = self._get_multiindex(dv)
+        dv_gb = self._groupby_multiindex(dv)
+        return dv_gb
+
+    def _drop_incomplete_djf(self, data_var: xr.DataArray) -> xr.DataArray:
+        """Drops months in incomplete DJF seasons.
+
+        Dropped months include the start year Jan/Feb and end year Dec.
+
+        Parameters
+        ----------
+        data_var : xr.DataArray
+            The data variable with some incomplete DJF seasons.
+
+        Returns
+        -------
+        xr.DataArray
+            The data variable with all complete DJF seasons.
+        """
+        start_year = data_var.time.dt.year.data[0]
+        end_year = data_var.time.dt.year.data[-1]
+
+        incomplete_seasons = (f"{start_year}-01", f"{start_year}-02", f"{end_year}-12")
+        for year_month in incomplete_seasons:
+            try:
+                data_var.loc[dict(time=year_month)] = np.nan
+                data_var = data_var.dropna("time")
+
+                self._time_bounds.loc[dict(time=year_month)] = np.nan
+                self._time_bounds = self._time_bounds.dropna("time")
```

Comment:
This probably doesn't affect performance, but I think these `.dropna("time")` calls could be placed outside the loop. Would something like this might make sense:

```
        try:
            for year_month in incomplete_seasons:
                data_var.loc[dict(time=year_month)] = np.nan
                self._time_bounds.loc[dict(time=year_month)] = np.nan
            data_var = data_var.dropna("time")
            self._time_bounds = self._time_bounds.dropna("time")
        except KeyError:
            logger.debug("Dataset does not contain {month}.")
            continue
```

Reply:
Nice optimization, better to perform the operation once if possible.

Reply:
The call to `.dropna("time")` assumes a continuous time series. Might need to address this by dropping specific coordinate points instead.

Reply:
The logic for this method has been updated to only drop the leading and trailing incomplete seasons.
https://github.com/XCDAT/xcdat/blob/b3e975f6d803a8314d4bd95ef5681c9c43794dec/xcdat/temporal_avg.py#L644-L661